### DISCUSSION TITLE
May 2023 release - 4.3 or 5.0

### DISCUSSION THREAD

**Discussion by gasman** [wagtail/wagtail#9980] [General]
With the first 4.2 release candidate now published, and the 4.2 branch in feature freeze, it's time for us to start planning the following release, which is scheduled for May 2023. One of the decisions to be made is whether this will be a 4.3 or 5.0 release. To cut to the chase - it's looking like 5.0 is the more attractive option, and I'd like to present the reasoning for that here.

As a reminder - [since we adopted semantic versioning](https://github.com/wagtail/wagtail/discussions/7732), the version number is not really a measure of the "size" of the release - 4.2 was pretty big in terms of new features, after all - but rather an indicator of how much potential impact and disruption an upgrade will have on existing projects, whether that's updating code in response to breaking changes, or briefing editorial users about significant changes to the user experience.

While we don't currently have any _new_ features in mind for the May release that would have that effect, one thing we'd want to do in a 5.0 release is drop all of the deprecated functionality that has been steadily accumulating since 3.0. Generally, when we deprecate something for future removal, we keep the old functionality working but have it output a deprecation warning, which means there's a lot of fallback code that has to remain in place until it's finally dropped once and for all. The full set of deprecations currently awaiting removal is as follows:

### Deprecated in 3.0:
Module renames from removal of wagtail.core
special-purpose field panels
use_json_field on StreamField
EditHandler internal API
ModelAdmin.get_edit_handler
ModelAdmin get_form_fields_exclude
BASE_URL -> WAGTAILADMIN_BASE_URL

### Deprecated in 4.0:
BaseSetting -> BaseSiteSetting
createFooChooser JS initialisers
PageRevision -> Revision method renames
wagtail.contrib.modeladmin.menus.SubMenu
base_url_path on ModelAdmin AdminURLHelper
AdminChooser -> BaseChooser
squashed testapp migrations
get_snippet_edit_handler

### Deprecated in 4.2:
Support for old Azure libraries in wagtail.contrib.frontendcache
page_locked_for_user -> locked_for_user
{% icon %} tag class_name

From this, we can see that the biggest changes, and the ones that most affect Wagtail sites as opposed to Wagtail extension packages, happened in 3.0 - the BaseSetting -> BaseSiteSetting change is probably the only later one affecting "normal" Wagtail sites. 3.0 was released in May 2022, so a 5.0 release in May 2023 would give site owners a full year to make those changes. Hopefully the vast majority have already done so, and so fully dropping these features won't cause any extra disruption beyond what site owners have already dealt with.

The `use_json_field` flag on StreamField is worth highlighting as a particular bugbear... the nature of that change means that we've had to roll it out in multiple steps to ensure that everyone has run the migration to JSONField before we remove support for TextField-based StreamFields, and in the interim period everyone has had to insert that bit of extra noise into their model definitions. The sooner 5.0 happens, the sooner we can move on to the final stage of dropping the need for that flag.

We always aim to give at least two releases between deprecating a feature and fully removing it, so under this plan, the things that have only just been deprecated in 4.2 will now be left in until version 6.0 (whenever that will be). Those items are all fairly minor, and it shouldn't be a big deal to keep their deprecation warnings in place for the next few releases. Since there were no deprecations in 4.1, developers have had at least 9 months warning of the things that are slated for removal in 5.0.

If there had been some more significant deprecations in 4.2 - for example, I was hoping to [bring all our search backends into agreement on how to handle autocompletion](https://github.com/wagtail/wagtail/issues/9903), but didn't have chance this time round - then there would have been a stronger case for delaying the 5.0 release until August, so that we could advance those 4.2 deprecations at the same time as the others. Since that wasn't the case, I think that a 5.0 release in May will give us the greatest possible value in terms of code cleanup, with the least possible disruption to users.

Happy to hear other thoughts!

**Comment by thibaudcolas** [comment]
👍 only thought for me is that the earlier we decide the better – so if 5.0 already appears more attractive, let’s go. One year for people to finish making the changes due to 3.0 deprecation does feel like plenty.

**Comment by andreobriennz** [comment]
Personally, I agree with the naming being 5.0. As a developer who has multiple sites to update, knowing how likely breaking changes are is quite important. And I've also heard other devs say they appreciate when a framework/CMS follows semantic versioning strictly.

But I also wonder if in the future breaking changes could be grouped together to be less frequent, similar to what Django does?

**Reply by gasman** [reply]
Thanks for the feedback! Given that breaking changes are going to arise at the same sort of rate whatever we do, there's a tradeoff between frequency and size of major releases - we don't want to let breaking changes pile up to the point that upgrading is a major hassle. I think it's best to make that decision on a case-by-case basis, rather than sticking to a fixed schedule.

**Comment by itzomen** [comment]
Sounds awesome, and we love the semantic versioning as it helps us communicate better with the stakeholders and plan for upgrades accordingly.

For example: Moving from 4-5 is gonna be a major tasks

@gasman My feature suggestion is a visual editor and if it's inline with Wagtail's roadmap I would love to contribute to the development. I've been exploring this since November as our clients are really excited in having that. Basically, we are thinking of leveraging GrapesJs to render the output editable on the preview section. I have a couple of ideas already but I'm lacking a lot of knowledge of Wagtail internals 😅.

Waiting for your feedback
Thanks
